### PR TITLE
feat: FlatPack environment

### LINKED ISSUE #187: Feature: Implement FlatPack environment
Implement the `FlatPack` environment including actor-critic and random agent networks.

### PR DESCRIPTION
**Pull request by RuanJohn**
### Details: 
Implements the full `FlatPack` environment with actor-critic and random networks as well as documentation. 
This environment was previously `Jigsaw`, the core changes from `Jigsaw` to `FlatPack` is that the blocks (previously pieces) do not have to be placed in exact places on the grid but instead the goal is to fill an empty grid with a given set of blocks. Please see the [previous closed PR](https://github.com/instadeepai/jumanji/pull/147) for comments that have been addressed. 

### Notes: 
Training of an a2c agent is complete and can results be viewed at the following [link](https://app.neptune.ai/o/InstaDeep/org/jumanji/runs/compare?viewId=998ef995-f828-46fc-960b-335b99b91dbf&detailsTab=metadata&shortId=JUJU-1381&dash=charts&type=run&compare=IwZg7AnANATFzyA&path=monitoring%2F&attribute=stderr).

closes #187

### CODE REVIEW

File: jumanji/environments/packing/flat_pack/types.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    blocks: 3D array with the blocks to be placed on the board. Here each block is a
+        2D array with shape (3, 3).
+    action_mask: 4D array showing where blocks can be placed on the grid.
+        this mask includes all possible rotations and possible placement locations
+        for each block on the grid.
+    """
+
+    current_grid: chex.Array  # (num_rows, num_cols)
+    blocks: chex.Array  # (num_blocks, 3, 3)
+    action_mask: chex.Array  # (num_blocks,)
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/types.py:37]:
```suggestion
    action_mask: chex.Array  # (num_blocks, num_rotations, num_rows-3, num_cols-3)
```

File: jumanji/environments/packing/flat_pack/utils.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    Args:
+        num_blocks: The number of blocks.
+    """
+    return 3 * num_blocks - (num_blocks - 1)
+
+
+def get_significant_idxs(grid_dim: int) -> chex.Array:
+    """Returns the indices of the grid that are significant. These will be used
+        to create interlocks between adjacent blocks.
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/utils.py:32]: nit
```suggestion
    """Returns the indices of the grid that are significant. These will be used
    to create interlocks between adjacent blocks.
```

File: jumanji/training/configs/config.yaml [hunk 3]
```diff
@@ -1,8 +1,8 @@
 defaults:
     - _self_
-    - env: snake  # [bin_pack, cleaner, connector, cvrp, game_2048, graph_coloring, job_shop, knapsack, maze, minesweeper, mmst, multi_cvrp, pac_man, robot_warehouse, rubiks_cube, snake, sokoban, sudoku, tetris, tsp]
+    - env: snake  # [bin_pack, cleaner, connector, cvrp, flat_packm game_2048, graph_coloring, job_shop, knapsack, maze, minesweeper, mmst, multi_cvrp, pac_man, robot_warehouse, rubiks_cube, snake, sokoban, sudoku, tetris, tsp]
```
Comment (sash-a) [inline jumanji/training/configs/config.yaml:3]:
```suggestion
    - env: snake  # [bin_pack, cleaner, connector, cvrp, flat_pack, game_2048, graph_coloring, job_shop, knapsack, maze, minesweeper, mmst, multi_cvrp, pac_man, robot_warehouse, rubiks_cube, snake, sokoban, sudoku, tetris, tsp]
```

File: jumanji/training/configs/config.yaml [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
 
-agent: random  # [random, a2c]
+agent: a2c  # [random, a2c]
```
Comment (sash-a) [inline jumanji/training/configs/config.yaml:5]:
```suggestion
agent: random  # [random, a2c]
```

File: jumanji/environments/packing/flat_pack/reward.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+
+class BlockDenseReward(RewardFn):
+    """Reward function for the dense reward setting.
+
+    This reward will give a normalised reward for each block placed on the grid
+        with each block being equally weighted. This implies that each placed block
+        will have a reward of `1 / num_blocks` and the maximum possible episode return
+        is 1. That is to say that, in the case of this reward, an agent will optimise
+        for placing as many blocks as possible on the grid.
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/reward.py:86]: nit
```suggestion
    This reward will give a normalised reward for each block placed on the grid
    with each block being equally weighted. This implies that each placed block
    will have a reward of `1 / num_blocks` and the maximum possible episode return
    is 1. That is to say that, in the case of this reward, an agent will optimise
    for placing as many blocks as possible on the grid.
```

File: jumanji/environments/packing/flat_pack/reward.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        next_state: State,
+        is_valid: bool,
+        is_done: bool,
+    ) -> chex.Numeric:
+        """Compute the reward based on the current state, the chosen action,
+        whether the action is valid and whether the episode is terminated.
+        """
+
+
+class CellDenseReward(RewardFn):
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/reward.py:39]: This seems like a bad reward to have, just rewarding based on how big your block is? So it'll bias the agent to place bigger blocks first? :thinking: 

I think a reward of `sum(state.grid != 0)` makes more sense

File: jumanji/environments/packing/flat_pack/types.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    key: random key used for board generation.
+    """
+
+    num_blocks: chex.Numeric  # ()
+    blocks: chex.Array  # (num_blocks, 3, 3)
+    action_mask: chex.Array  # (num_blocks, num_rotations, num_rows-3, num_cols-3)
+    placed_blocks: chex.Array  # (num_blocks,)
+    current_grid: chex.Array  # (num_rows, num_cols)
+    step_count: chex.Numeric  # ()
+    key: chex.PRNGKey  # (2,)
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/types.py:61]: More in line with other algorithms, have `current_grid` at the top and call it `grid`

Also seems to me that num_blocks and action_mask probably shouldn't be in the state

```suggestion
    grid: chex.Array  # (num_rows, num_cols)
    num_blocks: chex.Numeric  # ()
    blocks: chex.Array  # (num_blocks, 3, 3)
    action_mask: chex.Array  # (num_blocks, num_rotations, num_rows-3, num_cols-3)
    placed_blocks: chex.Array  # (num_blocks,)
    step_count: chex.Numeric  # ()
    key: chex.PRNGKey  # (2,)
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Initial review, looking good so far. Still need to review env, generator and training

File: jumanji/environments/packing/flat_pack/generator.py [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+        num_row_blocks: int,
+        num_col_blocks: int,
+    ):
+        """Initialises a random flat_pack generator.
+        Args:
+
+            num_row_blocks: Number of row blocks in flat_pack grid.
+            num_col_blocks: Number of column blocks in flat_pack grid.
+        """
+        super().__init__(num_row_blocks, num_col_blocks)
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/generator.py:76]: Same as the default so you can just leave this out
```suggestion
```

File: jumanji/environments/packing/flat_pack/generator.py [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+class ToyFlatPackGeneratorNoRotation(InstanceGenerator):
+    """Generates a deterministic toy FlatPack environment with 4 blocks. The
+    blocks are not rotated and not shuffled.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(
+            num_row_blocks=2,
+            num_col_blocks=2,
+        )
```
Comment (sash-a) [inline jumanji/environments/packing/flat_pack/generator.py:367]: nit
```suggestion
        super().__init__(num_row_blocks=2, num_col_blocks=2)
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Really great work @RuanJohn, mostly minor style things :smile:

File: jumanji/environments/packing/flat_pack/env.py [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+    def _get_ones_like_expanded_block(self, grid_block: chex.Array) -> chex.Array:
+        """Makes a grid of zeroes with ones where the block is placed.
+
+        Args:
+            grid_with_ones: block placed on a grid of zeroes.
+        """
+
+        grid_with_ones = jnp.where(grid_block != 0, 1, 0)
+
+        return grid_with_ones
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py]: Done

File: jumanji/environments/packing/flat_pack/generator.py (hunk 8 above)
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/generator.py]: Done

File: jumanji/environments/packing/flat_pack/generator.py [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+
+        num_blocks = self.num_row_blocks * self.num_col_blocks
+
+        # Compute the size of the grid.
+        grid_row_dim = compute_grid_dim(self.num_row_blocks)
+        grid_col_dim = compute_grid_dim(self.num_col_blocks)
+
+        # Get indices of grid where interlocks will be.
+        row_nibs_idxs = get_significant_idxs(grid_row_dim)
+        col_nibs_idxs = get_significant_idxs(grid_col_dim)
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/generator.py]: Done

File: jumanji/environments/packing/flat_pack/env.py [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+            lambda: state.current_grid + grid_block,
+            lambda: state.current_grid,
+        )
+        placed_blocks = jax.lax.cond(
+            action_is_legal,
+            lambda: state.placed_blocks.at[block_idx].set(True),
+            lambda: state.placed_blocks,
+        )
+
+        new_action_mask = self._make_action_mask(new_grid, state.blocks, placed_blocks)
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:200]: The action mask is include in the state of the `maze`, `cleaner` and `job_shop` environments and this is where I got the idea from. My suggestion is to leave it as is.

File: jumanji/environments/packing/flat_pack/env.py [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+
+        Returns:
+            True if the action is legal, False otherwise.
+        """
+
+        block_idx, _, _, _ = action
+
+        placed_mask = (current_grid > 0.0) + grid_mask_block
+
+        legal: bool = (~placed_blocks[block_idx]) & (jnp.max(placed_mask) <= 1)
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:376]: The first condition makes sure that a block hasn't been place on top of another block. So together the two conditions check that the same block isn't placed twice and that a blocks never overlap.

File: jumanji/environments/packing/flat_pack/env.py [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+            row_coord: row coordinate on the grid where the top left corner
+                of the block will be placed.
+            col_coord: column coordinate on the grid where the top left corner
+                of the block will be placed.
+
+        Returns:
+            Grid of zeroes with values where the block is placed.
+        """
+
+        grid_with_block = jnp.zeros((self.num_rows, self.num_cols), dtype=jnp.float32)
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:405]: Yes indeed, but at the end we are returning `grid_with_block` via a dynamic update slice. I've added a comment to try make it a bit more clear.

File: jumanji/environments/packing/flat_pack/reward.py (hunk 6 above)
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py:39]: I think this is not too bad though? And agents seem to be able to learn with it (it is the default reward). I think your suggestion could be its own reward since it would cumulative over time if I understand correctly.

File: jumanji/environments/packing/flat_pack/env.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (clement-bonnet) [reply jumanji/environments/packing/flat_pack/env.py:200]: I guess the action mask is needed to create the observation. But also, in the environments you mentioned above, the action mask is usually used to directly compute/read the value of `action_is_legal`. If possible, could we do that as well instead of recomputing the method `self._is_legal_action`?

File: jumanji/environments/packing/flat_pack/reward.py (hunk 6 above)
Comment (clement-bonnet) [reply jumanji/environments/packing/flat_pack/reward.py:39]: I personnally like this reward definition, summing to 1 if you complete the whole puzzle and favors more-completed puzzles than less-completed puzzles. Indeed, if discounted, this reward favors bigger blocks over smaller ones, but that's more of an agent problem (choose a discount not too low) since the environment is episodic anyway.

**Review by clement-bonnet — COMMENTED** [review]
We mustn't forget to update the docs as well (mkdocs.yaml)

File: jumanji/environments/packing/flat_pack/env.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:200]: This is quite a nice suggestion, I think we can do something like: 
```
action_is_legal = state.action_mask[block_idx, rotation, row_idx, col_idx]
```
Since the state will carry the action mask from the previous step. Let me see.

File: jumanji/environments/packing/flat_pack/reward.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        reward = jax.lax.cond(
+            is_valid,
+            lambda: jnp.float32(1.0 / num_blocks),
+            lambda: jnp.float32(0.0),
+        )
+
+        return reward
+
+
+class SparseReward(RewardFn):
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py:120]: Very nice suggestion! I do agree that the current one is very very sparse. But I do not feel like it is critical to implement another reward at the moment since our training curves are on the dense reward. Keen to hear your thoughts?

File: jumanji/environments/packing/flat_pack/reward.py (hunk 6 above)
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py:39]: I agree with this. Shall we then go with keeping this reward as is for now and then keep it as the default?

File: jumanji/environments/packing/flat_pack/reward.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py:120]: That is a fair point. I think we can just remove it in that case.

[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py:120]: Removed it :)

File: jumanji/environments/packing/flat_pack/reward.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        action: chex.Numeric,
+        next_state: State,
+        is_valid: bool,
+        is_done: bool,
+    ) -> chex.Numeric:
+        """Compute the reward based on the current state, the chosen action,
+        whether the action is valid and whether the episode is terminated.
+
+        Note here, that the action taken is not the raw action received from the
+        agent, but the block the agent opted to place on the grid.
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/reward.py]: Renamed `action` -> `placed_block` :)

File: jumanji/environments/packing/flat_pack/env.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:197]: Indeed, this works perfectly :)

**Review by sash-a — APPROVED** [review]
Looks pretty much good to go from my side. Happy with the action mask changes! Thanks @RuanJohn

**Review by clement-bonnet — DISMISSED** [review]
LGTM! 🔥 
I just have a few comments regarding doc and types. There needs to be a good reason to go from int32 to float32 when things are inherently integers. What would be the reason here?

File: jumanji/environments/packing/flat_pack/env.py [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+    ```
+    """
+
+    def __init__(
+        self,
+        generator: Optional[InstanceGenerator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        """Initializes the environment.
```
[the remark this answers is not in the document]
Comment (RuanJohn) [reply jumanji/environments/packing/flat_pack/env.py:106]:
```suggestion
        """Initializes the FlatPack environment.
```

**Review by clement-bonnet — APPROVED** [review]
LGTM
